src/HotChocolate/Caching/src/Caching/CacheControlTypeInterceptor.cs

```diff
@@ -0,0 +1,96 @@
+using System.Collections.Generic;
+using System.Linq;
+using HotChocolate.Configuration;
+using HotChocolate.Language;
+using HotChocolate.Types;
+using HotChocolate.Types.Descriptors;
+using HotChocolate.Types.Descriptors.Definitions;
+
+namespace HotChocolate.Caching;
+
+internal sealed class CacheControlTypeInterceptor : TypeInterceptor
+{
+    private readonly ICacheControlOptions _cacheControlOptions;
+
+    public CacheControlTypeInterceptor(ICacheControlOptionsAccessor accessor)
+    {
+        _cacheControlOptions = accessor.CacheControl;
+    }
+
+    public override void OnBeforeCompleteType(ITypeCompletionContext completionContext,
+        DefinitionBase definition)
+    {
+        if (!_cacheControlOptions.Enable || !_cacheControlOptions.ApplyDefaults)
+        {
+            return;
+        }
+
+        if (completionContext.IsIntrospectionType ||
+            completionContext.IsSubscriptionType == true ||
+            completionContext.IsMutationType == true)
+        {
+            return;
+        }
+
+        if (definition is not ObjectTypeDefinition objectDef)
+        {
+            return;
+        }
+
+        foreach (var field in objectDef.Fields)
+        {
+            if (field.IsIntrospectionField)
+            {
+                // Introspection fields do not need to be declared as cachable.
+                continue;
+            }
+
+            if (HasCacheControlDirective(field))
+            {
+                // If the field has a @cacheControl directive,
+                // we do not need to apply any defaults.
+                continue;
+            }
+
+            if (completionContext.IsQueryType == true ||
+                CostTypeInterceptor.IsDataResolver(field))
```

Comment:
@michaelstaib If I remember correctly you also wanted to move this `IsDataResolver` out of the Type Interceptor as a general utility.

Reply:
I will leave it there for now ... in general it would be better if we computed this when we configured the field ... but I do not want to touch this at the moment.